File: Sudoku/solver.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <time.h>
#include <math.h>
#include <limits.h>

#define N 9 // Size of sudoku NxN
#define SQRT_N 3 // Square root of N
#define MAX_RUNS 250
#define MIN_COST (N * N * 3) // each row, column and block should have cost N 
#define FILEPATH "Sudoku_instances/march_22_2025.txt"
#define NAME "march_22_2025"

typedef struct {
    int row;
    int col;
} Cell;

typedef struct {
    Cell cells[N];
    int size;
} Subgraph;

typedef struct {
    Subgraph rows[N];
    Subgraph columns[N];
    Subgraph blocks[N];
} Subgraphs;

int sudoku[N][N];
int cell_matrix[N][N];
int numbers[N];
int tour[N][N];
Subgraphs subgraphs;

void parse_sudoku_file(const char *file_path);
void save_sudoku_file(double time);
void generate_cell_matrix();
void generate_subgraphs();
bool validate_sudoku();
int calculate_global_cost(int tour[N][N]);
void generate_greedy_tour(Cell specific_cell, int specific_number);
void two_opt(int tour[N][N], Cell *cells, int n, int *global_cost);
void two_opt_reverse(int tour[N][N], Cell *cells, int n);
int two_opt_and_swap(int initial_tour[N][N], Cell *cells, int n);
void solve_sudoku(int runs);
void print_tour(int tour[N][N]);

void parse_sudoku_file(const char *file_path) {
    FILE *file = fopen(file_path, "r");
    if (!file) {
        printf("Error opening file %s\n", file_path);
        exit(1);
    }

    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            if (fscanf(file, "%d", &sudoku[i][j]) != 1) {
                printf("Error reading Sudoku from file %s\n", file_path);
                fclose(file);
                exit(1);
            }
        }
    }
    fclose(file);
}

void save_sudoku_file(double time) {
    char filepath[256];
    snprintf(filepath, sizeof(filepath), "Sudoku_results/%s_solution.txt", NAME);
    FILE *file = fopen(filepath, "w");
    if (!file) {
        printf("Error creating file %s\n", filepath);
        exit(1);
    }

    for (int r = 0; r < N; r++) {
        for (int c = 0; c < subgraphs.rows[r].size; c++) {
            Cell cell = subgraphs.rows[r].cells[c];
            fprintf(file, "%d ", tour[cell.row][cell.col]);
        }
        fprintf(file, "\n");
    }
    fprintf(file, "\ntotal time %.2f seconds", time);
    fclose(file);
    printf("Result saved in %s\n", filepath);
}

void generate_cell_matrix() {
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            cell_matrix[i][j] = sudoku[i][j];
        }
    }
}

void generate_subgraphs() {
    for (int i = 0; i < N; i++) {
        subgraphs.rows[i].size = N;
        for (int j = 0; j < N; j++) {
            subgraphs.rows[i].cells[j].row = i;
            subgraphs.rows[i].cells[j].col = j;
        }
    }

    for (int j = 0; j < N; j++) {
        subgraphs.columns[j].size = N;
        for (int i = 0; i < N; i++) {
            subgraphs.columns[j].cells[i].row = i;
            subgraphs.columns[j].cells[i].col = j;
        }
    }

    int block_idx = 0;
    for (int i = 0; i < N; i += SQRT_N) {
        for (int j = 0; j < N; j += SQRT_N) {
            subgraphs.blocks[block_idx].size = N;
            int idx = 0;
            for (int i2 = i; i2 < i + SQRT_N; i2++) {
                for (int j2 = j; j2 < j + SQRT_N; j2++) {
                    subgraphs.blocks[block_idx].cells[idx].row = i2;
                    subgraphs.blocks[block_idx].cells[idx].col = j2;
                    idx++;
                }
            }
            block_idx++;
        }
    }
}

bool validate_sudoku() {
    for (int s = 0; s < 3; s++) {
        Subgraph *structs = (s == 0) ? subgraphs.rows : (s == 1) ? subgraphs.columns : subgraphs.blocks;
        for (int i = 0; i < N; i++) {
            int used[N + 1] = {0};
            for (int j = 0; j < structs[i].size; j++) {
                Cell cell = structs[i].cells[j];
                int val = sudoku[cell.row][cell.col];
                if (val != 0) {
                    used[val]++;
                }
            }
            printf("{");
            bool first = true;
            for (int num = 1; num <= N; num++) {
                if (used[num] > 0) {
                    if (!first) printf(", ");
                    printf("%d: %d", num, used[num]);
                    first = false;
                }
            }
            printf("}\n");
            for (int num = 1; num <= N; num++) {
                if (used[num] > 1) {
                    printf("Invalid sudoku\n");
                    return false;
                }
            }
        }
    }
    printf("Sudoku seems valid.\n");
    return true;
}

int calculate_global_cost(int tour[N][N]) {
    int global_cost = 0;

    for (int s = 0; s < 3; s++) {
        Subgraph *structs = (s == 0) ? subgraphs.rows : (s == 1) ? subgraphs.columns : subgraphs.blocks;
        for (int i = 0; i < N; i++) {
            bool used[N + 1] = {false};
            int unique_count = 0;
            for (int j = 0; j < structs[i].size; j++) {
                Cell cell = structs[i].cells[j];
                int val = tour[cell.row][cell.col];
                if (!used[val]) {
                    used[val] = true;
                    unique_count++;
                }
            }
            global_cost += unique_count + (N - unique_count) * 2;
        }
    }
    return global_cost;
}

void generate_greedy_tour(Cell specific_cell, int specific_number) {
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            tour[i][j] = 0;
        }
    }

    for (int r = 0; r < N; r++) {
        bool used[N + 1] = {false};
        for (int c = 0; c < subgraphs.rows[r].size; c++) {
            Cell cell = subgraphs.rows[r].cells[c];
            if (cell_matrix[cell.row][cell.col] != 0 && !used[cell_matrix[cell.row][cell.col]]) {
                tour[cell.row][cell.col] = cell_matrix[cell.row][cell.col];
                used[cell_matrix[cell.row][cell.col]] = true;
            }
        }
        if (subgraphs.rows[r].cells[0].row == specific_cell.row && !used[specific_number]) {
            tour[specific_cell.row][specific_cell.col] = specific_number;
            used[specific_number] = true;
        }
        for (int c = 0; c < subgraphs.rows[r].size; c++) {
            Cell cell = subgraphs.rows[r].cells[c];
            if (tour[cell.row][cell.col] == 0) {
                for (int num = 1; num <= N; num++) {
                    if (!used[num]) {
                        tour[cell.row][cell.col] = num;
                        used[num] = true;
                        break;
                    }
                }
            }
        }
    }
    print_tour(tour);
}

void print_tour(int tour[N][N]) {
    printf("{");
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            printf("(%d,%d): %d", i, j, tour[i][j]);
            if (i < N - 1 || j < N - 1) printf(", ");
        }
    }
    printf("}\n");
}

void two_opt(int tour[N][N], Cell *cells, int n, int *global_cost) {
    *global_cost = calculate_global_cost(tour);
    bool improved = true;

    while (improved) {
        improved = false;
        for (int i = 0; i < n - 1; i++) {
            for (int j = i + 1; j < n; j++) {
                Cell cell1 = cells[i];
                Cell cell2 = cells[j];
                if (cell_matrix[cell1.row][cell1.col] == 0 && cell_matrix[cell2.row][cell2.col] == 0 &&
                    tour[cell1.row][cell1.col] != tour[cell2.row][cell2.col]) {
                    int temp = tour[cell1.row][cell1.col];
                    tour[cell1.row][cell1.col] = tour[cell2.row][cell2.col];
                    tour[cell2.row][cell2.col] = temp;
                    int new_cost = calculate_global_cost(tour);
                    if (new_cost < *global_cost) {
                        *global_cost = new_cost;
                        improved = true;
                    } else {
                        tour[cell2.row][cell2.col] = tour[cell1.row][cell1.col];
                        tour[cell1.row][cell1.col] = temp;
                    }
                }
            }
        }
    }
}

void two_opt_reverse(int tour[N][N], Cell *cells, int n) {
    int global_cost = calculate_global_cost(tour);
    bool improved = true;

    while (improved) {
        improved = false;
        for (int i = 0; i < n - 1; i++) {
            for (int j = i + 1; j < n; j++) {
                Cell cell1 = cells[i];
                Cell cell2 = cells[j];
                if (cell_matrix[cell1.row][cell1.col] == 0 && cell_matrix[cell2.row][cell2.col] == 0 &&
                    tour[cell1.row][cell1.col] != tour[cell2.row][cell2.col]) {
                    int temp = tour[cell1.row][cell1.col];
                    tour[cell1.row][cell1.col] = tour[cell2.row][cell2.col];
                    tour[cell2.row][cell2.col] = temp;
                    int new_cost = calculate_global_cost(tour);
                    if (new_cost > global_cost) {
                        global_cost = new_cost;
                        improved = true;
                    } else {
                        tour[cell2.row][cell2.col] = tour[cell1.row][cell1.col];
                        tour[cell1.row][cell1.col] = temp;
                    }
                }
            }
        }
    }
}

// For decision problems (NP-complete) generating a new local minimum from a pertubed local minimum have better results.
int two_opt_and_swap(int initial_tour[N][N], Cell *cells, int n) {
    int best_tour[N][N];
    int current_tour[N][N];
    int global_cost;

    memcpy(best_tour, initial_tour, sizeof(int) * N * N);
    two_opt_reverse(best_tour, cells, n);
    two_opt(best_tour, cells, n, &global_cost);
    memcpy(current_tour, best_tour, sizeof(int) * N * N);

    if (global_cost == MIN_COST) {
        printf("Found!\n");
        memcpy(tour, best_tour, sizeof(int) * N * N);
        return global_cost;
    }

    bool improved = true;
    while (improved) {
        improved = false;
        for (int i = 0; i < n - 1; i++) {
            for (int j = 0; j < n; j++) {
                if (i != j) {
                    Cell cell1 = cells[i];
                    Cell cell2 = cells[j];
                    if (cell_matrix[cell1.row][cell1.col] == 0 && cell_matrix[cell2.row][cell2.col] == 0 &&
                        current_tour[cell1.row][cell1.col] != current_tour[cell2.row][cell2.col]) {
                        int temp = current_tour[cell1.row][cell1.col];
                        current_tour[cell1.row][cell1.col] = current_tour[cell2.row][cell2.col];
                        current_tour[cell2.row][cell2.col] = temp;
                        int new_cost;
                        two_opt(current_tour, cells, n, &new_cost);
                        if (new_cost < global_cost) {
                            memcpy(best_tour, current_tour, sizeof(int) * N * N);
                            global_cost = new_cost;
                            improved = true;
                            printf("%d\n", global_cost);
                            if (global_cost == MIN_COST) {
                                printf("Found!\n");
                                memcpy(tour, best_tour, sizeof(int) * N * N);
                                return global_cost;
                            }
                            break;
                        }
                    }
                }
            }
            if (improved) break;
        }
    }
    print_tour(best_tour);
    memcpy(tour, best_tour, sizeof(int) * N * N);
    return global_cost;
}

void solve_sudoku(int runs) {
    clock_t start = clock();
    Cell cell_variations[MAX_RUNS];
    int number_variations[MAX_RUNS];
    int counter = 0;

    for (int i = 0; i < N && counter < runs; i++) {
        for (int j = 0; j < N && counter < runs; j++) {
            if (cell_matrix[i][j] == 0) {
                for (int num = 1; num <= N && counter < runs; num++) {
                    cell_variations[counter].row = i;
                    cell_variations[counter].col = j;
                    number_variations[counter] = num;
                    counter++;
                }
            }
        }
    }

    int lowest_cost = INT_MAX;
    Cell cells[N * N];
    int n = 0;
    for (int r = 0; r < N; r++) {
        for (int c = 0; c < subgraphs.rows[r].size; c++) {
            cells[n] = subgraphs.rows[r].cells[c];
            n++;
        }
    }

    for (int run = 0; run < counter; run++) {
        printf("Current run: %d\n", run + 1);
        printf("%.2f seconds\n", (double)(clock() - start) / CLOCKS_PER_SEC);
        printf("(%d,%d) %d\n", cell_variations[run].row, cell_variations[run].col, number_variations[run]);
        generate_greedy_tour(cell_variations[run], number_variations[run]);
        int cost = two_opt_and_swap(tour, cells, n);
        if (cost < lowest_cost) {
            lowest_cost = cost;
        }
        if (cost == MIN_COST) {
            break;
        }
    }

    printf("Saving result...\n");
    save_sudoku_file((double)(clock() - start) / CLOCKS_PER_SEC);
    printf("Saved.\n");
    printf("Total time: %.2f seconds\n", (double)(clock() - start) / CLOCKS_PER_SEC);
    printf("Lowest cost: %d\n", lowest_cost);
}

int main() {
    parse_sudoku_file(FILEPATH);
    for (int i = 0; i < N; i++) {
        numbers[i] = i + 1;
    }
    generate_cell_matrix();
    generate_subgraphs();
    if (!validate_sudoku()) {
        return 1;
    }
    solve_sudoku(MAX_RUNS);
    return 0;
}

File: Sudoku/solver.py
import time
import math

# Global variables
FILEPATH = "Sudoku_instances/april_5_2025.txt"
NAME = "april_5_2025"
MAX_RUNS = 250


def parse_sudoku_file(filepath):
    rows = []

    with open(filepath, 'r') as file:
        lines = file.readlines()
        file.close()

    for line in lines:
        numbers = line.split()
        row = []

        if len(numbers) == 0:
            break

        for number in numbers:
            row.append(int(number))
        rows.append(row)

    return rows


def save_sudoku_file(tour, subgraphs, time):
    filepath = f"Sudoku_results/{NAME}_solution.txt"
    rows = subgraphs[0]

    with open(filepath, 'w') as f:
        for row in rows:
            for cell in row:
                f.write(f"{tour[cell]} ")
            f.write("\n")

        f.write(f"\ntotal time {time} seconds")
        f.close()

    print(f"Result saved in {filepath}")


def generate_cell_matrix(sudoku):
    cell_matrix = {}
    n = len(sudoku[0])

    for i in range(n):
        for j in range(n):
            cell_matrix[(i, j)] = sudoku[i][j]

    return cell_matrix


def generate_subgraphs(numbers):
    n = len(numbers)
    sqrt = int(math.sqrt(n))

    rows = []
    columns = []
    blocks = []

    # Rows
    for i in range(n):
        row = set()

        for j in range(n):
            row.add((i, j))

        rows.append(row)

    # Columns
    for i in range(n):
        column = set()

        for j in range(n):
            column.add((j, i))

        columns.append(column)

    # Blocks
    for i in range(0, n, sqrt):
        for j in range(0, n, sqrt):
            block = set()

            for i2 in range(i, i + sqrt):
                for j2 in range(j, j + sqrt):
                    block.add((i2, j2))

            blocks.append(block)

    return [rows, columns, blocks]


def calculate_global_cost(tour, subgraphs):
    global_cost = 0
    n = len(subgraphs[0])

    # Rows
    for row in subgraphs[0]:
        unique_elements = set()

        for cell in row:
            unique_elements.add(tour[cell])

        global_cost += len(unique_elements) + (n - len(unique_elements)) * 2

    # Columns
    for column in subgraphs[1]:
        unique_elements = set()

        for cell in column:
            unique_elements.add(tour[cell])

        global_cost += len(unique_elements) + (n - len(unique_elements)) * 2

    # Blocks
    for block in subgraphs[2]:
        unique_elements = set()

        for cell in block:
            unique_elements.add(tour[cell])

        global_cost += len(unique_elements) + (n - len(unique_elements)) * 2

    return global_cost


def generate_greedy_tour(cell_matrix, numbers, specific_cell, specific_number, subgraphs):
    tour = {}

    for cell in cell_matrix:
        tour[cell] = 0

    for row in subgraphs[0]:
        available_numbers = set(numbers)

        for cell in row:
            if cell_matrix[cell] != 0 and cell_matrix[cell] in available_numbers:
                tour[cell] = cell_matrix[cell]
                available_numbers.remove(cell_matrix[cell])

        if specific_cell in row and specific_number in available_numbers:
            tour[specific_cell] = specific_number
            available_numbers.remove(specific_number)

        for cell in row:
            if tour[cell] == 0:
                for number in available_numbers:
                    tour[cell] = number
                    break

                available_numbers.remove(number)

    print(tour)

    return tour


# Adapted for Sudoku
def two_opt(tour, numbers, subgraphs, cell_matrix, cells):
    global_cost = calculate_global_cost(tour, subgraphs)
    improved = True

    n = len(cells)

    while improved:
        improved = False

        for i in range(n - 1):
            for j in range(i + 1, n):
                cell1 = cells[i]
                cell2 = cells[j]

                if cell_matrix[cell1] == 0 and cell_matrix[cell2] == 0:
                    if tour[cell1] != tour[cell2]:
                        temp = tour[cell1]
                        tour[cell1] = tour[cell2]
                        tour[cell2] = temp

                        new_global_cost = calculate_global_cost(tour, subgraphs)

                        if new_global_cost < global_cost:
                            global_cost = new_global_cost
                            improved = True
                        else:
                            tour[cell2] = tour[cell1]
                            tour[cell1] = temp

    return tour, global_cost


# Adapted for Sudoku
def two_opt_reverse(tour, numbers, subgraphs, cell_matrix, cells):
    global_cost = calculate_global_cost(tour, subgraphs)
    improved = True

    n = len(cells)

    while improved:
        improved = False

        for i in range(n - 1):
            for j in range(i + 1, n):
                cell1 = cells[i]
                cell2 = cells[j]

                if cell_matrix[cell1] == 0 and cell_matrix[cell2] == 0:
                    if tour[cell1] != tour[cell2]:
                        temp = tour[cell1]
                        tour[cell1] = tour[cell2]
                        tour[cell2] = temp

                        new_global_cost = calculate_global_cost(tour, subgraphs)

                        if new_global_cost > global_cost:
                            global_cost = new_global_cost
                            improved = True
                        else:
                            tour[cell2] = tour[cell1]
                            tour[cell1] = temp

    return tour


def two_opt_and_swap(initial_tour, numbers, subgraphs, cell_matrix):
    cells = []

    for row in subgraphs[0]:
        for cell in row:
            cells.append(cell)

    worsened_tour = two_opt_reverse(initial_tour, numbers, subgraphs, cell_matrix, cells)
    best_tour, global_cost = two_opt(worsened_tour, numbers, subgraphs, cell_matrix, cells)
    current_tour = best_tour.copy()
    min_cost = len(numbers) ** 2 * 3
    n = len(cells)

    improved = True

    if global_cost == min_cost:
        print("Found!")
        return best_tour, min_cost

    while improved:
        improved = False

        for i in range(n - 1):
            for j in range(n):
                if i != j:
                    cell1 = cells[i]
                    cell2 = cells[j]

                    if cell_matrix[cell1] == 0 and cell_matrix[cell2] == 0:
                        if current_tour[cell1] != current_tour[cell2]:
                            temp = current_tour[cell1]
                            current_tour[cell1] = current_tour[cell2]
                            current_tour[cell2] = temp
                            current_tour, new_global_cost = two_opt(current_tour, numbers, subgraphs, cell_matrix, cells)

                            if new_global_cost < global_cost:
                                best_tour = current_tour
                                global_cost = new_global_cost
                                improved = True
                                print(global_cost)

                                if global_cost == min_cost:
                                    print("Found!")
                                    return best_tour, global_cost
                                break
            if improved:
                break

    print(best_tour)
    return best_tour, global_cost


def validate_sudoku(sudoku, subgraphs):
    for structs in subgraphs:
        for struct in structs:
            used_numbers = {}

            for cell in struct:
                i = cell[0]
                j = cell[-1]

                if sudoku[i][j] != 0:
                    if sudoku[i][j] not in used_numbers:
                        used_numbers[sudoku[i][j]] = 1
                    else:
                        used_numbers[sudoku[i][j]] += 1

            print(used_numbers)

            for number in used_numbers:
                if used_numbers[number] > 1:
                    print("Invalid sudoku")
                    return False

    print("Sudoku seems valid.")
    return True


def solve_sudoku(sudoku, numbers, subgraphs):
    cell_matrix = generate_cell_matrix(sudoku)

    start = time.time()

    counter = 0
    run_counter = 0
    cell_variations = []

    min_cost = len(numbers) ** 2 * 3
    lowest_cost = float("inf")

    for cell in cell_matrix:
        if cell_matrix[cell] == 0:
            for number in numbers:
                cell_variations.append([cell, number])

                counter += 1
                if counter >= MAX_RUNS:
                    break
            if counter >= MAX_RUNS:
                break

    for cell_number in cell_variations:
        run_counter += 1
        print(f"Current run: {run_counter}")
        print(f"{round(time.time() - start, 2)} seconds")
        print(cell_number[0], cell_number[1])

        initial_tour = generate_greedy_tour(cell_matrix, numbers, cell_number[0], cell_number[1], subgraphs)
        tour, cost = two_opt_and_swap(initial_tour, numbers, subgraphs, cell_matrix)

        if cost < lowest_cost:
            lowest_cost = cost

        if cost == min_cost:
            break

    print("Saving result...")
    save_sudoku_file(tour, subgraphs, round(time.time() - start, 2))
    print("Saved.")

    print(f"Total time: {round(time.time() - start, 2)} seconds")
    print(f"Lowest cost: {lowest_cost}")


def main():
    sudoku = parse_sudoku_file(FILEPATH)
    numbers = list(range(1, len(sudoku) + 1))
    subgraphs = generate_subgraphs(numbers)

    is_valid = validate_sudoku(sudoku, subgraphs)
    if is_valid is False:
        return

    solve_sudoku(sudoku, numbers, subgraphs)


if __name__ == "__main__":
    main()
